File: Message/src/main/java/com/yahoo/sdvornik/message/codec/ByteBufToMsgDecoder.java
package com.yahoo.sdvornik.message.codec;

import com.yahoo.sdvornik.message.DataMessageBuffer;
import com.yahoo.sdvornik.message.Message;
import com.yahoo.sdvornik.message.StartSortingMessage;
import io.netty.buffer.ByteBuf;
import io.netty.channel.ChannelHandlerContext;
import io.netty.handler.codec.MessageToMessageCodec;
import io.netty.handler.codec.MessageToMessageDecoder;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.List;

public class ByteBufToMsgDecoder extends MessageToMessageDecoder<ByteBuf> {

    private static final Logger log = LoggerFactory.getLogger(ByteBufToMsgDecoder.class.getName());


    @Override
    protected void decode(ChannelHandlerContext ctx, ByteBuf in, List<Object> out) throws Exception {

        int typeInt = in.readInt();

        Message.Type type = Message.toEnum(typeInt);
        switch(type) {
            case GET_CONNECTION :
            case JOB_ENDED:
            case TASK_TRANSMISSION_ENDED:
            case GET_RESULT:
            case CONNECTED:
                out.add(Message.getSimpleOutboundMessage(type));
                break;

            case DATA_TYPE:
                int chunkNumber = in.readInt();
                out.add(new DataMessageBuffer(in, chunkNumber));
                in.retain();
                break;

            case START_SORTING:
                int taskCounter = in.readInt();
                out.add(new StartSortingMessage(taskCounter));
                break;
        }
    }
}

File: shared/src/main/kotlin/com/yahoo/sdvornik/message/codec/ByteBufToMsgDecoder.kt
package com.yahoo.sdvornik.message.codec

import com.yahoo.sdvornik.message.DataMessageBuffer
import com.yahoo.sdvornik.message.Message
import com.yahoo.sdvornik.message.StartSortingMessage
import io.netty.buffer.ByteBuf
import io.netty.channel.ChannelHandlerContext
import io.netty.handler.codec.MessageToMessageDecoder

/**
 * @author Serg Dvornik <sdvornik@yahoo.com>
 */
class ByteBufToMsgDecoder : MessageToMessageDecoder<ByteBuf>() {

    @Throws(Exception::class)
    override fun decode(ctx: ChannelHandlerContext, `in`: ByteBuf, out: MutableList<Any>) {

        val typeInt = `in`.readInt()

        val type = Message.toEnum(typeInt)

        when (type) {
            Message.Type.GET_CONNECTION,
            Message.Type.JOB_ENDED,
            Message.Type.TASK_TRANSMISSION_ENDED,
            Message.Type.GET_RESULT,
            Message.Type.CONNECTED -> out.add(Message.getSimpleOutboundMessage(type))

            Message.Type.DATA_TYPE -> {
                val chunkNumber = `in`.readInt()
                out.add(DataMessageBuffer(`in`, chunkNumber))
                `in`.retain()
            }

            Message.Type.START_SORTING -> {
                val taskCounter = `in`.readInt()
                out.add(StartSortingMessage(taskCounter))
            }
        }
    }
}